frontend/i_types.py

```diff
@@ -0,0 +1,274 @@
+from abc import ABCMeta, abstractmethod
+
+class Type(metaclass=ABCMeta):
+    """This class is the top-parent of all possible inferred types of a Python program.
+    Every type has its own class that inherits this class.
+    """
+
+    @abstractmethod
+    def is_subtype(self, t):
+        """Check if this type is a subtype of the parameter type.
+
+        Arguments:
+            type (Type): The type to check the subtype relationship against.
+        """
+        pass
+
+
+    @abstractmethod
+    def get_name(self):
+        """Get the name of the type"""
+        pass
+
+    # TODO: Implement __eq__ and __hash__ for all types, for correct set behaviour
+
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return True
+
+    def get_name(self):
+        return "None"
+
+
+class TBool(Type):
```

Comment:
IMO all of these (TBool, TInt, TFloat etc.) should be special cases of TClass. You could have special classes for them as you do now, but IMO they should then extend TClass. Then you wouldn't have to special-case is_subtype for those cases, you could handle it the same way as you do for all other classes.